src/utils/generator.js

```diff
@@ -52,12 +55,159 @@ import type {
+  | "ABSTRACT_OBJECT_GET_PROTO_OF"
+  | "ABSTRACT_OBJECT_GET"
+  | "DIRECT_CALL_WITH_ARG_LIST"
+  | "CALL_ABSTRACT_FUNC"
+  | "CALL_ABSTRACT_FUNC_THIS"
+  | "CALL_BAILOUT"
+  | "EMIT_CALL"
+  | "EMIT_CALL_AND_CAPTURE_RESULT"
+  | "GET_BINDING"
+  | "LOCAL_ASSIGNMENT"
+  | "LOGICAL_PROPERTY_ASSIGNMENT"
+  | "CONDITIONAL_PROPERTY_ASSIGNMENT"
+  | "PROPERTY_ASSIGNMENT"
+  | "MODULES_REQUIRE"
+  | "RESIDUAL_CALL"
+  | "ASSUME_CALL"
+  | "CANNOT_BECOME_OBJECT"
+  | "UPDATE_INCREMENTOR"
+  | "WIDENED_IDENTIFIER"
+  | "WIDEN_PROPERTY"
+  | "WIDEN_ABSTRACT_PROPERTY"
+  | "WIDEN_PROPERTY_ASSIGNMENT"
+  | "WIDEN_ABSTRACT_PROPERTY_ASSIGNMENT"
+  | "INVARIANT"
+  | "INVARIANT_APPEND"
+  | "DERIVED_ABSTRACT_INVARIANT"
+  | "PROPERTY_INVARIANT"
+  | "FULL_INVARIANT"
+  | "FULL_INVARIANT_FUNCTION"
+  | "FULL_INVARIANT_ABSTRACT"
+  | "UNKNOWN_ARRAY_METHOD_CALL"
+  | "UNKNOWN_ARRAY_METHOD_PROPERTY_CALL"
+  | "UNKNOWN_ARRAY_LENGTH"
+  | "UNKNOWN_ARRAY_GET_PARTIAL"
+  | "BABEL_HELPERS_OBJECT_WITHOUT_PROPERTIES"
+  | "REACT_DEFAULT_PROPS_HELPER"
+  | "REACT_TEMPORAL_FUNC"
+  | "REACT_CREATE_CONTEXT_PROVIDER"
+  | "REACT_SSR_REGEX_CONSTANT"
+  | "REACT_SSR_PREV_TEXT_NODE"
+  | "REACT_SSR_RENDER_VALUE_HELPER"
+  | "REACT_SSR_TEMPLATE_LITERAL"
+  | "REACT_NATIVE_STRING_LITERAL"
+  | "REACT_RELAY_MOCK_CONTAINER"
+  | "FB_MOCKS_BOOTLOADER_LOAD_MODULES"
+  | "FB_MOCKS_MAGIC_GLOBAL_FUNCTION";
+
+export type DerivedExpressionBuildNodeFunction = (
+  Array<BabelNodeExpression>,
+  SerializationContext,
+  Set<AbstractValue | ObjectValue>
+) => BabelNodeExpression;
+
+export type OperationDescriptor = {
+  data: OperationDescriptorData,
+  kind: void | ResidualBuildKind,
+  type: OperationDescriptorType,
+};
+
+// TODO: gradually remove all these, currently it's a random bag of values
+// that should be in args or in other places rather than here.
+export type OperationDescriptorData = {
+  appendLastToInvariantOperationDescriptor?: OperationDescriptor,
+  binding?: Binding | PropertyBinding,
+  boundName?: BabelNodeIdentifier,
+  callTemplate?: () => BabelNodeExpression,
+  concreteComparisons?: Array<Value>,
+  desc?: Descriptor,
+  generator?: Generator,
+  generators?: Array<Generator>,
+  id?: string,
+  lh?: BabelNodeVariableDeclaration,
+  op?: any, // TODO: This is a union of Babel operators, refactor to not use "any" at some point
+  prefix?: boolean,
+  path?: Value,
+  propertyGetter?: SupportedGraphQLGetters,
+  propName?: string,
+  propRef?: ReferenceName | AbstractValue,
+  object?: ObjectValue,
+  quasis?: Array<any>,
+  state?: "MISSING" | "PRESENT" | "DEFINED",
+  thisArg?: BaseValue | Value,
+  template?: PreludeGenerator => ({}) => BabelNodeExpression,
+  typeComparisons?: Set<typeof Value>,
+  typeofString?: string,
+  usesThis?: boolean,
+  value?: Value,
+  violationConditionOperationDescriptor?: OperationDescriptor,
+};
+
+export type ResidualBuildKind = "DERIVED" | "VOID";
+
+export function createOperationDescriptor(
```

Comment:
You can type check this instead of passing a random bag of values like this:

```js
type OperationsMap = {
  A: number,
  B: string,
};

function create<K: $Keys<OperationsMap>>(type: K, data: $ElementType<OperationsMap, K>) {
  return;
}

create('A', 42); // Ok
create('B', 42); // Error: `B` expects string data.
create('C', true); // Error: `C` is not a operation type.
```

[try-Flow](https://flow.org/try/#0C4TwDgpgBA8pBOBDYBLA9gOwM4FlFigF4oBvAKCigEEAuKDAVwFsAjCeAGgqgCE6tg8FBgDmXAL4BuMmQBmDDAGNUmKIvgRkEADwBpOgBJdEEFm1x2ydNjxgAfHYAUoSHV0coAE2SJDAUQAbCCYIDGAAFXAdCyQVG3wPXTsASlJuDWAGeAxpcRl1TWAIRwByKhKPABYAJmTJKAB6BtgAazICrVKeCqgausbmv3h4NHg6AAMecagIAA9IZSwoASFRLx8AOnaNTpKAYR7BBgh+pqghkbGocb3plCWMNGAoRCg0BCtVFwgtoA)

Unless you want to completely get rid of these eventually?

We could/should do a similar thing for abstract value `args`.

Reply:
Yeah the idea is to remove this object and move the data into args. We pass around functions and references to Prepack internals in some cases, so those particular operations will need refactoring quite heavily, but it's all incremental work.